File: src/questions/codesignal/firstduplicate/firstDuplicate.js
// const withObj = (arr) => {};

// const withSet = (arr) => {};

// Time Complexity: O(n) time or Linear Time
// Space Complexity: O(n) time or Linear Space
function firstDuplicate(a) {
  // intialize a empty object
  // let dict = {};

  // loop through the array
  // check whether the value is already in the object. if not add to it
  /*
  for (let i = 0; i < a.length; i++) {
    // check for presence of the element in object
    if (dict[a[i]]) {
      return a[i];
    }

    // add the item and set its value to true
    dict[a[i]] = true;
  }
  */

  // alternate solution using `Set`.
  // Preferred because there is no need for value in key/value pair of objects.
  let set = new Set();

  // loops and check each item of the array
  for (const item of a) {
    // returns item when set already has the same item.
    // space complexity for retrieving item is O(1)
    if (set.has(item)) {
      return item;
    }

    // add the item to the set for next loop
    set.add(item);
  }

  // value not found
  return -1;
}

let a = [2, 1, 3, 5, 3, 2];
console.log(firstDuplicate(a));

a = [2, 2];
console.log(firstDuplicate(a));

a = [2, 4, 3, 5, 1];
console.log(firstDuplicate(a));

File: src/questions/codesignal/firstduplicate/FirstDuplicate.java
package questions.codesignal.firstduplicate;

import java.util.HashSet;
import java.util.Set;

public class FirstDuplicate {

    // Time Complexity: O(1) time or Constant Time
    // Space Complexity: O(n) space or Linear Space
    public static int firstDuplicate(int[] a) {
        // initialize hash set. since it does not hold duplicates
        Set<Integer> set = new HashSet<>();

        // loop through the array and add each element to the set
        for (int j : a) {
            // when the element cannot be added to the set the duplicate value is found
            if (!set.add(j)) {
                return j;
            }
        }

        // there are no duplicates
        return -1;
    }

    public static void main(String[] args) {

        int[] a = {2, 1, 3, 5, 3, 2};
        System.out.println(firstDuplicate(a));

        int[] b = {2, 2};
        System.out.println(firstDuplicate(b));

        int[] c = {2, 4, 3, 5, 1};
        System.out.println(firstDuplicate(c));

    }

}
